File: LearningLinearAlgebra.FSharp.Tests/PolarRepresentationTests.fs
namespace LearningLinearAlgebra.Numbers

open Xunit

module PolarRepresentationTests =

    open LearningLinearAlgebra
    open LearningLinearAlgebra.Numbers
    open RealNumber
    open PolarPresentation

    let tolerance = 0.0000000001

    [<Fact>]
    let ``Sum of two complex numbers is calculated using the cartesian format`` () =
        let a = C(1, 0)
        let b = C(1, Pi / 2.0)

        let (sumMagniture, sumPhase) = Complex.Add a b |> Complex.Round |> Components

        sumMagniture |> Should.BeEquivalentTo.Real.Number(sqrt 2.0 |> Round)
        sumPhase |> Should.BeEquivalentTo.Real.Number(Pi / 4.0 |> Round)

    [<Fact>]
    let ``Difference of two complex numbers is calculated using the cartesian format`` () =
        let a = C(1, 0)
        let b = C(1, Pi / 2.0)

        let (differenceMagniture, differencePhase) =
            Complex.Subtract a b |> Complex.Round |> Components

        differenceMagniture |> Should.BeEquivalentTo.Real.Number(sqrt 2.0 |> Round)
        differencePhase |> Should.BeEquivalentTo.Real.Number(-1.0 * Pi / 4.0 |> Round)

    [<Fact>]
    let ``Product of two complex numbers is calculated by multiplying the magnitudes and adding their phase`` () =
        let a = C(sqrt 2.0, Pi / 4.0)
        let b = C(sqrt 2.0, Pi * (3.0 / 4.0))

        let (productMagniture, productPhase) =
            Complex.Multiply a b |> Complex.Round |> Components

        productMagniture |> Should.BeEquivalentTo.Real.Number(2)
        productPhase |> Should.BeEquivalentTo.Real.Number(Pi |> Round)

    [<Fact>]
    let ``Quotient of two complex numbers is calculated by dividing the magnitudes and subtracting their phase`` () =
        let a = C(sqrt 2.0, Pi / 4.0)
        let b = C(sqrt 2.0, Pi * (3.0 / 4.0))

        let (quotientMagniture, quotientPhase) =
            Complex.Divide a b |> Complex.Round |> Components

        quotientMagniture |> Should.BeEquivalentTo.Real.Number(1)
        quotientPhase |> Should.BeEquivalentTo.Real.Number((Pi * 1.5) |> Round)

    [<Fact>]
    let ``Another example of dividing complex numbers`` () =
        let a = C(sqrt (10.0), atan (-3.0))
        let b = C(sqrt (17.0), atan (4.0))

        let (quotientMagniture, quotientPhase) =
            Complex.Divide a b |> Complex.Round |> Components

        quotientMagniture |> Should.BeEquivalentTo.Real.Number(0.766965)
        quotientPhase |> Should.BeEquivalentTo.Real.Number(3.708322)

File: LearningLinearAlgebra.Tests/Numbers/PolarRepresentationTests.cs
using FluentAssertions;
using Xunit;
using static Computation.Numbers.RealNumber<float>;
using static Computation.Numbers.Polar<float>;

namespace LearningLinearAlgebra.Tests.Numbers;

public class PolarRepresentationTests
{
    [Fact]
    public void Sum_of_two_complex_numbers_is_calculated_using_the_cartesian_format()
    {
        var a = P(1, 0);
        var b = P(1, Pi / 2.0f);

        var sum = Add(a, b);

        sum.Magnitude.Should().Be(Sqrt(2));
        sum.Phase.Should().Be(Pi / 4.0f);
        sum.Should().Be(a + b);
    }

    [Fact]
    public void Difference_of_two_complex_numbers_is_calculated_using_the_cartesian_format()
    {
        var a = P(1, 0);
        var b = P(1, Pi / 2.0f);

        var difference = Subtract(a, b);

        difference.Magnitude.Should().Be(Sqrt(2));
        difference.Phase.Should().Be(Pi / -4.0f);
        difference.Should().Be(a - b);
    }

    [Fact]
    public void Product_of_two_complex_numbers_is_calculated_by_multiplying_the_magnitudes_and_adding_their_phase()
    {
        var a = P(Sqrt(2), Pi / 4.0f);
        var b = P(Sqrt(2), Pi * (3.0f / 4.0f));

        var product = Multiply(a, b);

        product.Magnitude.Should().BeApproximately(2, 10);
        product.Phase.Should().BeApproximately(Pi, 10);
        product.Should().Be(a * b);
    }

    [Fact]
    public void Quotient_of_two_complex_numbers_is_calculated_by_dividing_the_magnitudes_and_subtracting_their_phase()
    {
        var a = P(Sqrt(2), Pi / 4.0f);
        var b = P(Sqrt(2), Pi * (3.0f / 4.0f));


        var quotient = Divide(a, b);

        quotient.Magnitude.Should().Be(1);
        quotient.Phase.Should().Be(Pi * 1.5f);
        quotient.Should().Be(a / b);
    }

    [Fact]
    public void Another_example_of_dividing_complex_numbers()
    {
        var a = P(Sqrt(10), Atan(-3.0f));
        var b = P(Sqrt(17), Atan(4.0f));

        var quotient = Divide(a, b);

        quotient.Magnitude.Should().BeApproximately(0.7669649888f, 10);
        quotient.Phase.Should().BeApproximately(3.7083218711f, 10);
        quotient.Should().Be(a / b);
    }
}